golang/vaas/pkg/vaas/vaas.go

```diff
@@ -0,0 +1,340 @@
+		}, err
+	}
+
+	subscription := v.broadcastChannel.Subscribe()
+	defer v.broadcastChannel.RemoveSubscription(subscription)
+
+	request := msg.NewVerdictRequest(v.sessionId, v.options, sha256)
+	v.requestChannel <- request
+
+	response, err := v.waitForResponse(subscription, request.GetGuid())
+	if err != nil {
+		return msg.VaasVerdict{
+			Verdict: msg.Verdict(msg.Error),
+			Sha256:  sha256,
+		}, err
+	}
+
+	if response.Verdict == msg.Verdict(msg.Unknown) {
+		if err := v.uploadFile(data, response.Url, response.UploadToken); err != nil {
+			return msg.VaasVerdict{
+				Verdict: msg.Verdict(msg.Unknown),
+				Sha256:  sha256,
+			}, err
+		} else {
+			uploadResponse, err := v.waitForResponse(subscription, request.GetGuid())
+			if err != nil {
+				return msg.VaasVerdict{
+					Verdict: msg.Verdict(msg.Error),
+					Sha256:  sha256,
+				}, err
+			} else {
+				response = uploadResponse
+			}
+		}
+	}
+
+	return msg.VaasVerdict{
+		Verdict: response.Verdict,
+		Sha256:  response.Sha256,
+	}, nil
+}
+
+func (v *vaas) ForFileList(fileList []string) ([]msg.VaasVerdict, error) {
+	if v.sessionId == "" {
+		return []msg.VaasVerdict{}, errors.New("invalid operation")
+	}
+
+	var waitGroup sync.WaitGroup
+	var verdicts []msg.VaasVerdict
+
+	for _, file := range fileList {
+		waitGroup.Add(1)
+
+		go func(file string) {
+			defer waitGroup.Done()
+			verdict, err := v.ForFile(file)
+			if err != nil {
+				verdicts = append(verdicts, msg.VaasVerdict{Sha256: verdict.Sha256, Verdict: msg.Verdict(msg.Error)})
+			} else {
+				verdicts = append(verdicts, verdict)
+			}
+		}(file)
+
+		waitGroup.Wait()
+	}
+	return verdicts, nil
+}
+
+func (v *vaas) ForUrl(url string) (msg.VaasVerdict, error) {
+	if v.sessionId == "" {
+		return msg.VaasVerdict{}, errors.New("invalid operation")
+	}
+	subscription := v.broadcastChannel.Subscribe()
+	defer v.broadcastChannel.RemoveSubscription(subscription)
+	request := msg.NewVerdictRequestForUrl(v.sessionId, v.options, url)
+	v.requestChannel <- request
+
+	verdictResponse, responseErr := v.waitForResponse(subscription, request.GetGuid())
+	if responseErr != nil {
+		return msg.VaasVerdict{
+			Verdict: msg.Verdict(msg.Error),
+		}, responseErr
+	}
+
+	return msg.VaasVerdict{
+		Verdict: verdictResponse.Verdict,
+		Sha256:  verdictResponse.Sha256,
+	}, nil
+}
+
+func (v *vaas) uploadFile(file *os.File, url string, token string) error {
+	httpClient := &http.Client{}
+	file.Seek(0, 0)
+	info, _ := file.Stat()
+	req, err := http.NewRequest(http.MethodPut, url, file)
+	if err != nil {
+		return err
+	}
+	req.Header.Add("Authorization", token)
+	req.ContentLength = int64(info.Size())
+
+	httpResponse, err := httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if httpResponse.StatusCode != 200 {
+		return errors.New("StatusCode:" + fmt.Sprintf("%x", httpResponse.StatusCode))
+	}
+
+	return nil
+}
+
+func (v *vaas) waitForResponse(subscription <-chan msg.VerdictResponse, guid string) (msg.VerdictResponse, error) {
```

Comment:
In my opinion it would be better to let the user pass a context in the ForXXX-functions to dertermine a timeout. You could still use the 180s as kind of global timeout but maybe a user has a use case, where he can't wait this long for a request or just want's to cancel it for any reasons. This is not possible with the current design.